Jetpack E2E: add test for the stats tracking pixel.

worldomonation opened the pull request:
Related to: https://github.com/Automattic/wp-calypso/issues/80730

## Proposed Changes

This PR adds coverage for the tracking pixel, used to count stats. 

## Testing Instructions

Ensure the following build configurations are passing:
  - [x] Calypso E2E (desktop)
  - [x] Jetpack E2E Simple (mobile)
  - [ ] Jetpack E2E Simple (desktop)
  - [x] Jetpack E2E AT (desktop)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

worldomonation committed:
test/e2e/specs/editor/editor__post-basic-flow.ts
- work in coverage for the tracking pixel.

worldomonation committed:
test/e2e/specs/editor/editor__post-basic-flow.ts
- do not visit the post as part of publishing step.
- open a new window to check the published post's state.

worldomonation commented on test/e2e/specs/editor/editor__post-basic-flow.ts:
```diff
@@ -160,16 +162,33 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
 			await editorPage.saveDraft();
 		} );
+
+		it( 'Publish post', async function () {
+			publishedURL = await editorPage.publish();
+		} );
```
Because we are no longer viewing the post as the publishing user, split this off into its own step and only do the publishing step.

worldomonation commented on test/e2e/specs/editor/editor__post-basic-flow.ts:
```diff
@@ -160,16 +162,33 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
+		it( 'View published post', async function () {
+			// Check for `blog` and `post` query params, used for stats tracking.
+			const trackingPixelLoaded = newPage.waitForResponse(
+				/pixel\.wp\.com\/g\.gif.*blog=[\d]+.*&post=[\d]+/
```
This regex is pretty specific about the 1) content and 2) structure of the request.

dpasque replied:
Nice, I think it looks great!

**Question (very non-blocking):** I believe we should be able to know both the blog _and_ the post ID at this point: any reason to not just look for them exactly in the regex? (As opposed to any other characters.) I could be wrong, but it seems like a nice extra level of validation that _shouldn't_ add any more flakiness!

worldomonation replied:
We don't actually know the post ID! `EditorPage.publish` doesn't return the intercepted response, but rather just the URL of the published post. While we can do some regex hackery to extract the post ID, I don't think that's a reliable solution.

We **do** know the blogID though, so that's something we can check against.

dpasque replied:
Awesome, seems like a great middle-ground to me! 🚀

worldomonation committed:
test/e2e/specs/editor/editor__post-basic-flow.ts
- remove private user from target

dpasque commented on test/e2e/specs/editor/editor__post-basic-flow.ts:
```diff
@@ -160,16 +162,34 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
+				/pixel\.wp\.com\/g\.gif.*blog=[\d]+.*&post=[\d]+/
+			);
+			await newPage.goto( publishedURL.href );
+			const response = await trackingPixelLoaded;
+			expect( response.status() ).toBe( 200 );
```
**Suggestion (non-blocking):** How do you feel about pulling this assertion into its own test step? That way if this fails, the test failure is clear about which specific part failed! That is, as opposed to it looking like we failed to navigate to the post. We can of course (and should) start the promise in this step, more just about resolving and checking it later in its own!

worldomonation replied:
The issue with starting the promise in a step after the actual navigation to the post, is that `waitForResponse` doesn't have visibility into requests that completed prior to calling this method. As [upstream documentation](https://playwright.dev/docs/network#variations) suggests, the `waitForResponse` has to wrap around the action triggering network requests.

Case in point, this is what happens if the `waitForResponse` call is moved into a new step after the `Visit Published Post` step. This is run on my local machine:

```
 page.waitForResponse: Timeout 10000ms exceeded while waiting for event "response"
    =========================== logs ===========================
    waiting for response /pixel.wp.com\/g.gif.*blog=212495662+.*&post=[\d]+/
    ============================================================
```

worldomonation replied:
I worked around the issue and updated the PR. 

The idea is to wrap the `waitForResponse` in a try/catch, assign the response to a holding value, and then check whether it's defined in the next step. This way if the tracking pixel ever fails to load, it will not fail at the "Visit Published Post" step but rather the "Jetpack Stats tracking pixel is loaded" step.

dpasque replied:
> As [upstream documentation](https://playwright.dev/docs/network#variations) suggests, the waitForResponse has to wrap around the action triggering network requests.

> Case in point, this is what happens if the waitForResponse call is moved into a new step after the Visit Published Post step.

Bah, I should have made myself clearer, sorry! I just meant passing the promise onto the next step instead of awaiting it in that one. I definitely think we need to keep it wrapped around.

So, something like...

```typescript
let trackingPixelLoadedPromise: Promise< Response >;
it( 'View published post', async function () {
  // Check for `blog` and `post` query params, used for stats tracking.
  trackingPixelLoadedPromise = newPage.waitForResponse(
  new RegExp(
	  `pixel.wp.com/g.gif.*blog=${ testAccount.credentials.testSites?.primary.id }+.*&post=[\\d]+`
  )
  );
  await newPage.goto( publishedURL.href );
 
  expect( publishedURL.href ).toStrictEqual( newPage.url() );
  } );
  
it( 'Jetpack Stats tracking pixel is loaded', async function () {
  const response = await trackingPixelLoadedPromise;
  expect( response ).toBeDefined();
  expect( response.status() ).toBe( 200 );
} );
```

But, what you have totally works too and is fine! 👍

worldomonation replied:
> So, something like...
> 
> 
> 
> ```typescript
> 
> let trackingPixelLoadedPromise: Promise< Response >;
> 
> it( 'View published post', async function () {
> 
>   // Check for `blog` and `post` query params, used for stats tracking.
> 
>   trackingPixelLoadedPromise = newPage.waitForResponse(
> 
>   new RegExp(
> 
> 	  `pixel.wp.com/g.gif.*blog=${ testAccount.credentials.testSites?.primary.id }+.*&post=[\\d]+`
> 
>   )
> 
>   );
> 
>   await newPage.goto( publishedURL.href );
> 
>  
> 
>   expect( publishedURL.href ).toStrictEqual( newPage.url() );
> 
>   } );
> 
>   
> 
> it( 'Jetpack Stats tracking pixel is loaded', async function () {
> 
>   const response = await trackingPixelLoadedPromise;
> 
>   expect( response ).toBeDefined();
> 
>   expect( response.status() ).toBe( 200 );
> 
> } );
> 
> ```

I actually did what you suggested as the first step, and it was what produced the error log that I pasted. 

It does align with previous experience re:waitForResponse behavior. The request must fire after the waitForResponse call starts, and since the request starts soon after the page is loaded the separated call ends up timing out.

dpasque reviewed (approved):
Fantastic test addition! 🚀  I added two small optional suggestions, but otherwise should be good-to-go!

worldomonation committed:
test/e2e/specs/editor/editor__post-basic-flow.ts
- move tracking pixel expect statement to a new step.
- specify blog ID in url regex.

worldomonation committed:
test/e2e/specs/editor/editor__post-basic-flow.ts
- wrap the waitForResponse in a try/catch and fail at the right step.

dpasque commented:
And feel free to merge whenever! 🚀
